# **Feature Engineering**

## Objectives

* Write your notebook objective here, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write down which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* If you have any additional comments that don't fit in the previous bullets, please state them here. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [124]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\sonia\\Documents\\VS Studio Projects\\US_Air_Pollution_Team_2'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [129]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


In [134]:
os.chdir(r"c:\Users\sonia\Documents\VS Studio Projects\US_Air_Pollution_Team_2")

os.getcwd()

'c:\\Users\\sonia\\Documents\\VS Studio Projects\\US_Air_Pollution_Team_2'

Confirm the new current directory

In [135]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\sonia\\Documents\\VS Studio Projects\\US_Air_Pollution_Team_2'

---

# Required Libraries

In [127]:
import pandas as pd
import numpy as np
import re
from datetime import datetime
from meteostat import Point, Daily

---

# Load the Dataset

I will load the dataset using Pandas and look at the first 5 rows.

In [136]:
df = pd.read_csv('Dataset/Processed/pollution_us_2012_2016-cleaned.csv') # Reading the CSV file
df # Displaying the first 5 rows of the dataframe

State Code  County Code  Site Num  \
0                4           13      3002   
1                4           13      3002   
2                4           13      3002   
3                4           13      3002   
4                4           13      3002   
...            ...          ...       ...   
136329          56           21       100   
136330          56           21       100   
136331          56           21       100   
136332          56           21       100   
136333          56           21       100   

                                        Address    State    County  \
0       1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa   
1       1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa   
2       1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa   
3       1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa   
4       1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa   
...                                         ...      ...       ...   
136329    NCore - North Cheyenne Soccer Complex  Wyoming   Laramie   
136330    NCore - North Cheyenne Soccer Complex  Wyoming   Laramie   
136331    NCore - North Cheyenne Soccer Complex  Wyoming   Laramie   
136332    NCore - North Cheyenne Soccer Complex  Wyoming   Laramie   
136333    NCore - North Cheyenne Soccer Complex  Wyoming   Laramie   

                 City  Date Local          NO2 Units   NO2 Mean  \
0             Phoenix  2012-01-01  Parts per billion  21.208332   
1             Phoenix  2012-01-02  Parts per billion  17.208332   
2             Phoenix  2012-01-03  Parts per billion  30.000000   
3             Phoenix  2012-01-04  Parts per billion  33.666668   
4             Phoenix  2012-01-05  Parts per billion  31.695652   
...               ...         ...                ...        ...   
136329  Not in a city  2016-03-27  Parts per billion   4.277273   
136330  Not in a city  2016-03-28  Parts per billion   8.317391   
136331  Not in a city  2016-03-29  Parts per billion   2.564706   
136332  Not in a city  2016-03-30  Parts per billion   1.083333   
136333  Not in a city  2016-03-31  Parts per billion   0.939130   

        NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI           O3 Units  \
0                    33.0                 0       31  Parts per million   
1                    38.0                22       36  Parts per million   
2                    47.0                18       44  Parts per million   
3                    47.0                19       44  Parts per million   
4                    48.0                18       45  Parts per million   
...                   ...               ...      ...                ...   
136329               23.5                23       22  Parts per million   
136330               22.6                 6       21  Parts per million   
136331                3.6                 6        3  Parts per million   
136332                1.6                 9        1  Parts per million   
136333                1.3                 5        1  Parts per million   

         O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  \
0       0.015083             0.028               11      24   
1       0.018042             0.034                9      29   
2       0.008542             0.024               10      20   
3       0.005458             0.016               10      14   
4       0.008292             0.024                9      20   
...          ...               ...              ...     ...   
136329  0.041958             0.050               10      46   
136330  0.041292             0.052                9      48   
136331  0.028000             0.040               23      37   
136332  0.043917             0.048               18      44   
136333  0.045263             0.047                9      44   

                SO2 Units  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  \
0       Parts per billion  1.458333                5.0                 0   
1       Par

---

## Keep Useful Columns

I'm going to make another DataFrame and keep the columns that I think will be useful. I'm doing this, rather than dropping coliumns, so that if we change our mind, it will be easier to remove/ add columns. 

I am going to remove codes related to location and the units for each of the measurements.

In [138]:
keep_col = ["Address",
            "State",
            "County",
            "City",
            "Date Local",
            "NO2 Mean",
            "NO2 1st Max Value",
            "NO2 1st Max Hour",
            "NO2 AQI",
            "O3 Mean",
            "O3 1st Max Value",
            "O3 1st Max Hour",
            "O3 AQI",
            "SO2 Mean",
            "SO2 1st Max Value",
            "SO2 1st Max Hour",
            "SO2 AQI",
            "CO Mean",
            "CO 1st Max Value", 
            "CO 1st Max Hour",
            "CO AQI",
]

df_keep = df[keep_col]
df_keep.head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0  2012-01-01  21.208332               33.0                 0       31   
1  2012-01-02  17.208332               38.0                22       36   
2  2012-01-03  30.000000               47.0                18       44   
3  2012-01-04  33.666668               47.0                19       44   
4  2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI  
0                5    31.0  
1                0     6.0  
2               23    19.0  
3                1    22.0  
4               23    19.0

I am looking for rows that now appear as duplicated, without the unkept columns.

In [139]:
df_keep[df_keep.duplicated()]

Address     State   County               City  \
1070           PIKE AVE AT RIVER ROAD  Arkansas  Pulaski  North Little Rock   
1072           PIKE AVE AT RIVER ROAD  Arkansas  Pulaski  North Little Rock   
1254           PIKE AVE AT RIVER ROAD  Arkansas  Pulaski  North Little Rock   
1256           PIKE AVE AT RIVER ROAD  Arkansas  Pulaski  North Little Rock   
1258           PIKE AVE AT RIVER ROAD  Arkansas  Pulaski  North Little Rock   
...                               ...       ...      ...                ...   
133142  10TH ST. & VINE ST. DAVENPORT      Iowa    Scott          Davenport   
133144  10TH ST. & VINE ST. DAVENPORT      Iowa    Scott          Davenport   
133146  10TH ST. & VINE ST. DAVENPORT      Iowa    Scott          Davenport   
133148  10TH ST. & VINE ST. DAVENPORT      Iowa    Scott          Davenport   
133170  10TH ST. & VINE ST. DAVENPORT      Iowa    Scott          Davenport   

        Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
1070    2012-02-02  18.183332               46.6                19       43   
1072    2012-02-02  18.183332               46.6                19       43   
1254    2012-03-19   5.082609                8.7                21        8   
1256    2012-03-19   5.082609                8.7                21        8   
1258    2012-03-20   6.483333               18.4                22       17   
...            ...        ...                ...               ...      ...   
133142  2016-05-17   6.790909               15.2                21       14   
133144  2016-05-18   7.136364               22.8                23       21   
133146  2016-05-19   7.177273               17.3                21       16   
133148  2016-05-20   9.336364               15.9                 7       14   
133170  2016-05-31   5.390909               11.5                 1       10   

         O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
1070    0.015667             0.027               10      23  1.320833   
1072    0.015667             0.027               10      23  1.154167   
1254    0.031208             0.039               13      33  1.408333   
1256    0.031208             0.039               13      33  0.962500   
1258    0.025208             0.032               10      27  1.220833   
...          ...               ...              ...     ...       ...   
133142  0.024625             0.033                7      31  0.034783   
133144  0.033167             0.050               11      46  0.013043   
133146  0.039000             0.056               11      54  0.165217   
133148  0.033375             0.051               11      47  0.434783   
133170  0.032895             0.042               10      39  0.082609   

        SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  \
1070                  3.1                10      4.0  0.700000   
1072                  2.8                10      3.0  0.700000   
1254                  2.4                19      3.0  0.400000   
1256                  1.9                19      1.0  0.400000   
1258                  2.1                11      3.0  0.400000   
...                   ...               ...      ...       ...   
133142                0.2                 0      0.0  0.250000   
133144                0.1                 7      0.0  0.195833   
133146                0.7                 8      0.0  0.229167   
133148                1.0                 7      1.0  0.225000   
133170                0.5                10      0.0  0.200000   

        CO 1st Max Value  CO 1st Max Hour  CO AQI  
1070                 0.8                0     9.0  
1072                 0.8                0     9.0  
1254                 0.4                0     5.0  
1256                 0.4                0     5.0  
1258                 0.4                0     5.0  
...                  ...              ...     ...  
133142               0.4               10     5.0  
133144               0.3               22     

There are 316 duplicated rows which I'll now drop.

In [140]:
df_keep = df_keep.drop_duplicates()
df_keep.shape

(136018, 21)

---

## Add New Column "Population"

I am going to add population data for city/ county, which is available from the US Census website. We would like to see how population relates to pollutant levels.

In [141]:
df_city_pop = pd.read_csv("Dataset/Support_files/City_Pop_Map.csv", encoding='latin1')
df_city_pop

Area Population
0       Abbeville city, Alabama      2,349
1      Adamsville city, Alabama      4,393
2         Addison town, Alabama        661
3           Akron town, Alabama        229
4       Alabaster city, Alabama     33,342
...                         ...        ...
21408                       NaN        NaN
21409                       NaN        NaN
21410                       NaN        NaN
21411                       NaN        NaN
21412                       NaN        NaN

[21413 rows x 2 columns]

I will get rid of the NaN from the bottom rows.

In [142]:
df_city_pop = df_city_pop.dropna(how='all')
df_city_pop

Area Population
0       Abbeville city, Alabama      2,349
1      Adamsville city, Alabama      4,393
2         Addison town, Alabama        661
3           Akron town, Alabama        229
4       Alabaster city, Alabama     33,342
...                         ...        ...
19474   Wamsutter town, Wyoming        203
19475   Wheatland town, Wyoming      3,586
19476     Worland city, Wyoming      4,784
19477      Wright town, Wyoming      1,645
19478       Yoder town, Wyoming        129

[19479 rows x 2 columns]

Next, I will separate the city from the state so that they are in separate columns, and drop the column "Area".

In [143]:
df_city_pop = df_city_pop.copy()

# Remove leading/trailing spaces 
df_city_pop['Area'] = df_city_pop['Area'].str.strip()

# Split at the last comma into city and state
df_city_pop[['City', 'State']] = df_city_pop['Area'].str.rsplit(',', n=1, expand=True)

# Strip extra spaces and remove "city" in the name
df_city_pop['City'] = df_city_pop['City'].str.replace(r'\b[Cc]ity\b', '', regex=True).str.strip()
df_city_pop['State'] = df_city_pop['State'].str.strip()

# Optional: drop 'Area'
df_city_pop = df_city_pop.drop(columns=['Area'])

df_city_pop.head()

Population          City    State
0      2,349     Abbeville  Alabama
1      4,393    Adamsville  Alabama
2        661  Addison town  Alabama
3        229    Akron town  Alabama
4     33,342     Alabaster  Alabama

To enhance the chances of matching the two dataframes on "City" and "State", I am going to create "City_norm" and "State_norm" columns in both datasets, and that no spaces or other symbols are present, all names are lower case, and the word "city" is removed.  

In [144]:
def normalize_city(name):
    name = name.lower().strip()
    
    # Remove parenthetical content, e.g., " (city)"
    name = re.sub(r"\s*\(.*?\)\s*", "", name)
    
    # Remove common suffixes
    remove_terms = [" city"]
    for term in remove_terms:
        name = name.replace(term, "")
    
    # Remove extra spaces
    name = re.sub(r"\s+", " ", name).strip()
    
    return name

# Apply to both datasets
df_keep['City_norm'] = df_keep['City'].apply(normalize_city)
df_keep['State_norm'] = df_keep['State'].str.lower().str.strip()

df_city_pop['City_norm'] = df_city_pop['City'].apply(normalize_city)
df_city_pop['State_norm'] = df_city_pop['State'].str.lower().str.strip()

Before merging, I would like to see how many cities that are in the pollution dataset (df_keep) are also in the population dataset (df_city_pop).

In [145]:
# Make sets of cities (and states) in each dataset
pol_cities = set(zip(df_keep['City'], df_keep['State']))
pop_cities = set(zip(df_city_pop['City'], df_city_pop['State']))

# Count how many cities in df_balanced are in df_pop
matched = pol_cities & pop_cities
print(f"Cities in df_keep found in df_pop: {len(matched)}")
print(f"Cities in df_keep NOT found in df_pop: {len(pol_cities - pop_cities)}")

Cities in df_keep found in df_pop: 63
Cities in df_keep NOT found in df_pop: 39


This shows that 63 cities are matched and 39 are missing. 

I will now read in the county level population dataset.

In [146]:
df_county_pop = pd.read_csv("Dataset/Support_files/County_Pop_Map.csv", encoding='latin1')
df_county_pop.head()

County Population
0  .Autauga County, Alabama     58,800
1  .Baldwin County, Alabama    231,767
2  .Barbour County, Alabama     25,226
3     .Bibb County, Alabama     22,284
4   .Blount County, Alabama     59,130

Once again, I will split the county and state into separate columns and strip and spaces.

In [147]:
df_county_pop = df_county_pop.copy()

# Step 1: Remove leading "."
df_county_pop["County"] = df_county_pop["County"].str.lstrip(".")

# Step 2: Split into "County" and "State"
df_county_pop[["County", "State"]] = df_county_pop["County"].str.replace(" County", "", regex=False).str.rsplit(",", n=1, expand=True)

# Step 3: Clean up whitespace
df_county_pop["County"] = df_county_pop["County"].str.strip()
df_county_pop["State"] = df_county_pop["State"].str.strip()

# Check result
df_county_pop.head()

County Population    State
0  Autauga     58,800  Alabama
1  Baldwin    231,767  Alabama
2  Barbour     25,226  Alabama
3     Bibb     22,284  Alabama
4   Blount     59,130  Alabama

I will create a "County_norm" column in both datasets, but only a "State_norm" column in the county population dataset (df_county_pop) because the pollution dataset already has this column.

In [148]:
def normalize_county(name):
    name = name.lower().strip()
    
    # Remove parenthetical content, e.g., " (city)"
    name = re.sub(r"\s*\(.*?\)\s*", "", name)
    
    # Replace common suffixes
    remove_terms = [
        " county", " parish", " borough", " census area", 
        " independent city", " municipality", " district",
        " planning region", " region"
    ]
    for term in remove_terms:
        name = name.replace(term, "")
    
    # Normalize "st." or "st" to "st"
    name = re.sub(r"\bst\.?", "st", name)
    
    # Remove extra spaces
    name = re.sub(r"\s+", " ", name).strip()
    
    return name

# Apply to dataset
df_keep['County_norm'] = df_keep['County'].apply(normalize_county)

df_county_pop['County_norm'] = df_county_pop['County'].apply(normalize_county)
df_county_pop['State_norm'] = df_county_pop['State'].str.lower().str.strip()

I would like to see how many counties that are in the pollution dataset are found in the county population dataset.

In [149]:
# Make sets of counties (and states) in each dataset
pol_counties = set(zip(df_keep['County_norm'], df_keep['State_norm']))
pop_counties = set(zip(df_county_pop['County_norm'], df_county_pop['State_norm']))

# Count how many cities in df_balanced are in df_pop
matched = pol_counties & pop_counties
print(f"Counties in df found in df_county_pop: {len(matched)}")
print(f"Counties in df NOT found in df_county_pop: {len(pol_counties - pop_counties)}")

Counties in df found in df_county_pop: 88
Counties in df NOT found in df_county_pop: 5


I can see that 88 counties were matched, while 5 were not. Below are a list of the unmatched counties.

In [150]:
list(pol_counties - pop_counties)[:5]

[('litchfield', 'connecticut'),
 ('hartford', 'connecticut'),
 ('fairfield', 'connecticut'),
 ('new haven', 'connecticut'),
 ('saint clair', 'illinois')]

Before merging, I am going to look for city-state and county-state pairs which have become identical due to normalisation.  

In [151]:
# Check df_pop for duplicate city-state pairs
dupes_city_pop = df_city_pop[df_city_pop.duplicated(subset=['City_norm', 'State_norm'], keep=False)]
print(f"Duplicate city-state pairs in df_city_pop: {dupes_city_pop.value_counts().sum()}")
print(dupes_city_pop.sort_values(['City_norm', 'State_norm']))

# Check df_cpop for duplicate county-state pairs
dupes_county_pop = df_county_pop[df_county_pop.duplicated(subset=['County_norm', 'State_norm'], keep=False)]
print(f"Duplicate county-state pairs in df_county_pop: {dupes_county_pop.value_counts().sum()}")
print(dupes_county_pop.sort_values(['County_norm', 'State_norm']))

Duplicate city-state pairs in df_city_pop: 49
      Population                   City         State              City_norm  \
3269       4,714        Beecher village      Illinois        beecher village   
3270         429       Beecher  village      Illinois        beecher village   
14706        177    Centerville borough  Pennsylvania    centerville borough   
14707      3,257    Centerville borough  Pennsylvania    centerville borough   
16718      2,853            Clarksville         Texas            clarksville   
16719        768            Clarksville         Texas            clarksville   
6686       1,030                   Clay      Kentucky                   clay   
6687       1,194                   Clay      Kentucky                   clay   
14740        129       Coaldale borough  Pennsylvania       coaldale borough   
14741      2,427       Coaldale borough  Pennsylvania       coaldale borough   
18976        234          Genoa village     Wisconsin          genoa villa

I can see that there are 49 city-state duplicates and 0 county-state duplicates. I am going to drop the city-state duplicates because they will cause a problem when merging.

In [152]:
df_city_pop_clean = df_city_pop[~df_city_pop.index.isin(dupes_city_pop.index)].copy()

Now, I will merge population data at city level first, and check the shape to esure no extra rows or columns have been added beyond what is expected.

In [153]:
# Merge city population
df_merge = df_keep.merge(df_city_pop_clean[['City_norm','State_norm','Population']],
                                left_on=['City_norm','State_norm'],
                                right_on=['City_norm','State_norm'],
                                how='left')
df_merge.rename(columns={'Population':'Population_city'}, inplace=True)

df_merge.shape

(136018, 25)

Next, I will merge population at county level and check the shape.

In [154]:
# Merge county population
df_merge = df_merge.merge(df_county_pop[['County_norm','State_norm','Population']],
                                left_on=['County_norm','State_norm'],
                                right_on=['County_norm','State_norm'],
                                how='left')
df_merge.rename(columns={'Population':'Population_county'}, inplace=True)

df_merge.shape

(136018, 26)

Lastly, I will merge available population data into a "Population" column, which first looks to fill from city level data, and then by county level if city level is not available.

In [155]:
# Fill final population using city first, then County
df_merge['Population'] = df_merge['Population_city'].fillna(df_merge['Population_county'])

df_merge.shape

(136018, 27)

I would like to summarise how many values are missing from each column.

In [156]:
print(df_merge["Population_city"].isna().sum())
print(df_merge["Population_county"].isna().sum())
print(df_merge["Population"].isna().sum())

37024
5111
3664


I can see that 3664 rows have missing data from the newly added "Population" column. I will drop these rows.

In [157]:
df_merge = df_merge.dropna(subset=['Population'])
df_merge.shape

(132354, 27)

I'll have a look at the new dataframe.

In [158]:
df_merge.head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0  2012-01-01  21.208332               33.0                 0       31   
1  2012-01-02  17.208332               38.0                22       36   
2  2012-01-03  30.000000               47.0                18       44   
3  2012-01-04  33.666668               47.0                19       44   
4  2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI City_norm State_norm County_norm Population_city  \
0                5    31.0   phoenix    arizona    maricopa       1,608,415   
1                0     6.0   phoenix    arizona    maricopa       1,608,415   
2               23    19.0   phoenix    arizona    maricopa       1,608,415   
3                1    22.0   phoenix    arizona    maricopa       1,608,415   
4               23    19.0   phoenix    arizona    maricopa       1,608,415   

  Population_county Population  
0         4,425,315  1,608,415  
1         4,425,315  1,608,415  
2         4,425,315  1,608,415  
3         4,425,315  1,608,415  
4         4,425,315  1,608,415

I will need to get rid of the comma in the "Population" column and change the data type to number.

In [ ]:
df['Population'] = df['Population'].str.replace(',', '').astype(int)

---

## Add Weather Conditions

I'd like to add basic weather data as environmental conditions are known to have an effect on the level of pollution. To do this, I am going to use the Meteostate library which holds historical weather data from various weather stations. 

The first step will be to get county centroids (latitude and longitude), which I will use to find the nearest weather station. County centroid data, which was found on kaggle, will be read in below. 

In [159]:
county_coords = pd.read_csv("Dataset/Support_files/County_Centroids.csv", encoding='utf-8-sig')
county_coords.head()

state          county  cfips  latitude  longitude
0  Alabama  Autauga County   1001   32.5081   -86.6513
1  Alabama  Baldwin County   1003   30.7725   -87.7842
2  Alabama  Barbour County   1005   31.8832   -85.3931
3  Alabama     Bibb County   1007   33.0388   -87.0967
4  Alabama   Blount County   1009   34.0126   -86.5339

I need to capitalise the names of the columns in the centroid dataset (county-coords) so that it will matache the pollution dataset.

In [160]:
# Clean column names
county_coords.columns = county_coords.columns.str.strip()  # remove weird characters like 
county_coords.rename(
    columns={
        "state": "State",
        "county": "County",
        "latitude": "Latitude",
        "longitude": "Longitude"
    },
    inplace=True
)

# Preview
county_coords.head()

State          County  cfips  Latitude  Longitude
0  Alabama  Autauga County   1001   32.5081   -86.6513
1  Alabama  Baldwin County   1003   30.7725   -87.7842
2  Alabama  Barbour County   1005   31.8832   -85.3931
3  Alabama     Bibb County   1007   33.0388   -87.0967
4  Alabama   Blount County   1009   34.0126   -86.5339

I will get rid of the unwanted column, "cfips".

In [161]:
county_coords = county_coords.drop(["cfips"], axis=1)
county_coords.head()

State          County  Latitude  Longitude
0  Alabama  Autauga County   32.5081   -86.6513
1  Alabama  Baldwin County   30.7725   -87.7842
2  Alabama  Barbour County   31.8832   -85.3931
3  Alabama     Bibb County   33.0388   -87.0967
4  Alabama   Blount County   34.0126   -86.5339

Then I create normalised columns, only on the county centroid dataset, because they are already present on the pollution dataset, so I can match the two datasets.

In [162]:
def normalize_county(name):
    name = name.lower().strip()
    
    # Remove parenthetical content, e.g., " (city)"
    name = re.sub(r"\s*\(.*?\)\s*", "", name)
    
    # Replace common suffixes
    remove_terms = [
        " county", " parish", " borough", " census area", 
        " independent city", " municipality", " district",
        " planning region", " region"
    ]
    for term in remove_terms:
        name = name.replace(term, "")
    
    # Normalize "st." or "st" to "st"
    name = re.sub(r"\bst\.?", "st", name)
    
    # Remove extra spaces
    name = re.sub(r"\s+", " ", name).strip()
    
    return name

# Apply to dataset
county_coords['County_norm'] = county_coords['County'].apply(normalize_county)
county_coords['State_norm'] = county_coords['State'].str.lower().str.strip()

county_coords.head()

State          County  Latitude  Longitude County_norm State_norm
0  Alabama  Autauga County   32.5081   -86.6513     autauga    alabama
1  Alabama  Baldwin County   30.7725   -87.7842     baldwin    alabama
2  Alabama  Barbour County   31.8832   -85.3931     barbour    alabama
3  Alabama     Bibb County   33.0388   -87.0967        bibb    alabama
4  Alabama   Blount County   34.0126   -86.5339      blount    alabama

Next, I'll check how many counties are matched between the two datasets.

In [163]:
# Check how many normalized (State, County) pairs match before merging
matches = df_merge.merge(
    county_coords[['State_norm', 'County_norm']],
    on=['State_norm', 'County_norm'],
    how='inner'
)

print(f"Number of matching counties: {len(matches)}")
print(f"Total counties in df_merge: {len(df_merge)}")
print(f"Match rate: {len(matches) / len(df_merge) * 100:.2f}%")

Number of matching counties: 132354
Total counties in df_merge: 132354
Match rate: 100.00%


I can see that all counties match, so I am ready to merge.

In [164]:
merged = df_merge.merge(
    county_coords[['State_norm', 'County_norm', 'Latitude', 'Longitude']],
    on=['State_norm', 'County_norm'],
    how='left'
)

merged.head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0  2012-01-01  21.208332               33.0                 0       31   
1  2012-01-02  17.208332               38.0                22       36   
2  2012-01-03  30.000000               47.0                18       44   
3  2012-01-04  33.666668               47.0                19       44   
4  2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI City_norm State_norm County_norm Population_city  \
0                5    31.0   phoenix    arizona    maricopa       1,608,415   
1                0     6.0   phoenix    arizona    maricopa       1,608,415   
2               23    19.0   phoenix    arizona    maricopa       1,608,415   
3                1    22.0   phoenix    arizona    maricopa       1,608,415   
4               23    19.0   phoenix    arizona    maricopa       1,608,415   

  Population_county Population  Latitude  Longitude  
0         4,425,315  1,608,415   33.2798  -112.7681  
1         4,425,315  1,608,415   33.2798  -112.7681  
2         4,425,315  1,608,415   33.2798  -112.7681  
3         4,425,315  1,608,415   33.2798  -112.7681  
4         4,425,315  1,608,415   33.2798  -112.7681

I can see that "Latitude" and "Longitude" columns have been added. I will double check all rows have a value.

In [165]:
print(merged["Latitude"].isna().sum())
print(merged["Latitude"].isna().sum())

0
0


## Addition of Available Weather Data

Now I am ready to look for weathe data using Meteostat. First, I am going to rename the dataset "df".

In [184]:
df = merged.copy()
df.head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0  2012-01-01  21.208332               33.0                 0       31   
1  2012-01-02  17.208332               38.0                22       36   
2  2012-01-03  30.000000               47.0                18       44   
3  2012-01-04  33.666668               47.0                19       44   
4  2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI City_norm State_norm County_norm Population_city  \
0                5    31.0   phoenix    arizona    maricopa       1,608,415   
1                0     6.0   phoenix    arizona    maricopa       1,608,415   
2               23    19.0   phoenix    arizona    maricopa       1,608,415   
3                1    22.0   phoenix    arizona    maricopa       1,608,415   
4               23    19.0   phoenix    arizona    maricopa       1,608,415   

  Population_county Population  Latitude  Longitude  
0         4,425,315  1,608,415   33.2798  -112.7681  
1         4,425,315  1,608,415   33.2798  -112.7681  
2         4,425,315  1,608,415   33.2798  -112.7681  
3         4,425,315  1,608,415   33.2798  -112.7681  
4         4,425,315  1,608,415   33.2798  -112.7681

I need to check the data type of "Date Local" as it will need to be datetime to be used to pull data.

In [185]:
df['Date Local'].dtype

dtype('O')

I will change it from object to datatime.

In [186]:
df['Date Local'] = pd.to_datetime(df['Date Local'])
df['Date Local'].dtype

dtype('<M8[ns]')

Here, I am going to look for weather data for each unique location in the dataset. First, we use "Latitude" and "Longitude" to find the nearest station, but if that fails, we use "County" and "State". "Date Local" is also used to get the required weather records.  

In [187]:
# Keep only unique locations
unique_locations = df[['State', 'County', 'Latitude', 'Longitude']].drop_duplicates()

weather_data = []

for _, loc in unique_locations.iterrows():
    try:
        # Create a Point for the location
        point = Point(loc['Latitude'], loc['Longitude'])
        
        # Determine date range for this location
        dates = df[(df['Latitude'] == loc['Latitude']) &
                    (df['Longitude'] == loc['Longitude'])]
        start = dates['Date Local'].min()
        end = dates['Date Local'].max()
        
        # Find nearest stations (up to 5) with data in the date range
        stations = Stations().nearby(loc['Latitude'], loc['Longitude'])
        stations = stations.fetch(15)
        
        # Pick the first station with data available for the date range
        station_found = False
        for station_id in stations.index:
            try:
                daily = Daily(station_id, start, end).fetch()
                if not daily.empty:
                    daily['State'] = loc['State']
                    daily['County'] = loc['County']
                    weather_data.append(daily)
                    station_found = True
                    break
            except:
                continue
        
        if not station_found:
            print(f"No data for {loc['County']}, {loc['State']} in the date range.")
    
    except Exception as e:
        print(f"Error processing {loc['County']}, {loc['State']}: {e}")

# Combine all locations
weather_df = pd.concat(weather_data).reset_index()

weather_df.head()

time  tavg  tmin  tmax  prcp  snow  wdir  wspd  wpgt  pres  tsun  \
0 2013-01-01   4.6  -2.0  12.0  <NA>  <NA>  <NA>   8.6  <NA>  <NA>  <NA>   
1 2013-01-02   7.8   1.0  16.0  <NA>  <NA>  <NA>   9.2  <NA>  <NA>  <NA>   
2 2013-01-03   9.1  <NA>  <NA>  <NA>  <NA>  <NA>  13.5  <NA>  <NA>  <NA>   
3 2013-01-04   8.8  <NA>  <NA>  <NA>  <NA>  <NA>  12.8  <NA>  <NA>  <NA>   
4 2013-01-05   6.0  -2.0  15.0  <NA>  <NA>  <NA>   9.3  <NA>  <NA>  <NA>   

     State    County  
0  Arizona  Maricopa  
1  Arizona  Maricopa  
2  Arizona  Maricopa  
3  Arizona  Maricopa  
4  Arizona  Maricopa

I can see that many values are NA. I will summarise the number of missing values for the columns I am interested in.

In [188]:
print(f"Missing tavg {weather_df["tavg"].isna().sum()}")
print(f"Missing tmax {weather_df["tmax"].isna().sum()}")
print(f"Missing wspd {weather_df["wspd"].isna().sum()}")
print(f"Missing prcp {weather_df["prcp"].isna().sum()}")

Missing tavg 7812
Missing tmax 4209
Missing wspd 5972
Missing prcp 40431


I was unsure whether to use average temperature "tavg" or maximum temperature "tmax", but I have decided to go with maximum temperature as more data is available.

I will merge the datasets to add the weather columns into the pollution dataset.

In [189]:
# Merge with original data to align exact dates
final_df = df.merge(
    weather_df,
    left_on=['State', 'County', 'Date Local'],
    right_on=['State', 'County', 'time'],
    how='left'
)

# Drop duplicate 'time' column if desired
final_df = final_df.drop(columns=['time'])
pd.set_option("display.max_columns", None)
final_df. head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

  Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0 2012-01-01  21.208332               33.0                 0       31   
1 2012-01-02  17.208332               38.0                22       36   
2 2012-01-03  30.000000               47.0                18       44   
3 2012-01-04  33.666668               47.0                19       44   
4 2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI City_norm State_norm County_norm Population_city  \
0                5    31.0   phoenix    arizona    maricopa       1,608,415   
1                0     6.0   phoenix    arizona    maricopa       1,608,415   
2               23    19.0   phoenix    arizona    maricopa       1,608,415   
3                1    22.0   phoenix    arizona    maricopa       1,608,415   
4               23    19.0   phoenix    arizona    maricopa       1,608,415   

  Population_county Population  Latitude  Longitude  tavg  tmin  tmax  prcp  \
0         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  <NA>   
1         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  <NA>   
2         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  <NA>   
3         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  <NA>   
4         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  <NA>   

   snow  wdir  wspd  wpgt  pres  tsun  
0  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  
1  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  
2  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  
3  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  
4  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>

I'm going to have a check that the date ranges match.

In [191]:
# For your main DataFrame
print("final_df date range:", final_df['Date Local'].min(), "to", final_df['Date Local'].max())

# For the weather DataFrame
print("weather_df date range:", weather_df['time'].min(), "to", weather_df['time'].max())

final_df date range: 2012-01-01 00:00:00 to 2016-05-31 00:00:00
weather_df date range: 2012-01-01 00:00:00 to 2016-05-31 00:00:00


Now I will drop the weather data columns that are not needed.

In [192]:
final_df = final_df.drop(columns=["tavg", "tmin", "snow", "wdir", "wpgt", "pres", "tsun"])
final_df.head()

Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

  Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0 2012-01-01  21.208332               33.0                 0       31   
1 2012-01-02  17.208332               38.0                22       36   
2 2012-01-03  30.000000               47.0                18       44   
3 2012-01-04  33.666668               47.0                19       44   
4 2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI City_norm State_norm County_norm Population_city  \
0                5    31.0   phoenix    arizona    maricopa       1,608,415   
1                0     6.0   phoenix    arizona    maricopa       1,608,415   
2               23    19.0   phoenix    arizona    maricopa       1,608,415   
3                1    22.0   phoenix    arizona    maricopa       1,608,415   
4               23    19.0   phoenix    arizona    maricopa       1,608,415   

  Population_county Population  Latitude  Longitude  tmax  prcp  wspd  
0         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  
1         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  
2         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  
3         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>  
4         4,425,315  1,608,415   33.2798  -112.7681  <NA>  <NA>  <NA>

I'd like to see how many rows have missing data for the weather columns.

In [193]:
print(f"Rows with missing tmax {final_df['tmax'].isna().sum()}")
print(f"Rows with missing wspd {final_df['wspd'].isna().sum()}")
print(f"Rows with missing prcp {final_df['prcp'].isna().sum()}")

Rows with missing tmax 15310
Rows with missing wspd 15009
Rows with missing prcp 57047


In [194]:
df_final_imputed = final_df.copy()

I can see that we have a lot of missing weather values. I'm going to try and fill some values by taking the previous days, but no more than the previous days, as I believe this will lead to too many artefacts in the data. I am also going to group by "Address", so that only weather data for that location is used to fill missing values.

In [198]:
weather_cols = ["tmax", "prcp", "wspd"]

df_final_imputed[weather_cols] = (
    df_final_imputed.groupby('Address')[weather_cols]
    .transform(lambda x: x.ffill(limit=1))
)

Below I'll summarise how many missing values remain.

In [200]:
print(f"Rows with missing tmax {df_final_imputed['tmax'].isna().sum()}")
print(f"Rows with missing wspd {df_final_imputed['wspd'].isna().sum()}")
print(f"Rows with missing prcp {df_final_imputed['prcp'].isna().sum()}")

Rows with missing tmax 10205
Rows with missing wspd 14070
Rows with missing prcp 51172


I can see that most missing values remain; I have filled ~12,000 values. For those missing values, this suggests that they are missing for the entire address, or at least, for the first row of the address. I am going to drop the rest of these rows with missing values. We should still have sufficient data to understand trends and perform ML.

To be safe, I will save two separate csv files, one with the weather data and reduced rows, and one without the weather data (and reduced row).

In [201]:
df_final_imputed = df_final_imputed.dropna()
print(df_final_imputed.shape)
df_final_imputed.head()

(58158, 32)


Address    State County    City Date Local   NO2 Mean  \
638  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-01  17.716667   
639  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-02  15.062500   
640  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-03  21.643478   
641  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-04  25.041668   
642  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-05  21.981817   

     NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI   O3 Mean  O3 1st Max Value  \
638               31.0                 0       29  0.013667             0.030   
639               30.6                18       28  0.015083             0.030   
640               31.0                18       29  0.011417             0.026   
641               37.8                10       35  0.009208             0.020   
642               37.1                17       35  0.013042             0.031   

     O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  \
638               10      25  0.254167                0.5                19   
639               10      25  0.200000                0.6                19   
640                9      22  0.295455                0.7                 8   
641               10      17  0.737500                2.1                19   
642                9      26  0.330435                0.8                21   

     SO2 AQI   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI City_norm  \
638      0.0  0.336842               0.6                5     7.0    tucson   
639      0.0  0.225000               0.4               23     5.0    tucson   
640      0.0  0.295833               0.4                0     5.0    tucson   
641      3.0  0.345833               0.5               12     6.0    tucson   
642      0.0  0.291667               0.6               23     7.0    tucson   

    State_norm County_norm Population_city Population_county Population  \
638    arizona        pima         542,649         1,043,441    542,649   
639    arizona        pima         542,649         1,043,441    542,649   
640    arizona        pima         542,649         1,043,441    542,649   
641    arizona        pima         542,649         1,043,441    542,649   
642    arizona        pima         542,649         1,043,441    542,649   

     Latitude  Longitude  tmax  prcp  wspd  
638   31.9681  -111.7806  26.7   0.0  17.6  
639   31.9681  -111.7806  24.4   0.0  27.4  
640   31.9681  -111.7806  26.1   0.0  10.8  
641   31.9681  -111.7806  24.4   0.0   9.0  
642   31.9681  -111.7806  23.9   0.0   9.7

Dropping rows with missing weather data has reduced the number of rows from 132354 to 58158.

---

## Save DataFrames to CSV Files

This will be the csv file with the population and weather columns added (58158 rows).

In [202]:
keep_col = ["Address",
            "State",
            "County",
            "City",
            "Date Local",
            "NO2 Mean",
            "NO2 1st Max Value",
            "NO2 1st Max Hour",
            "NO2 AQI",
            "O3 Mean",
            "O3 1st Max Value",
            "O3 1st Max Hour",
            "O3 AQI",
            "SO2 Mean",
            "SO2 1st Max Value",
            "SO2 1st Max Hour",
            "SO2 AQI",
            "CO Mean",
            "CO 1st Max Value", 
            "CO 1st Max Hour",
            "CO AQI",
            "Population",
            "Latitude",
            "Longitude",
            "tmax",
            "prcp",
            "wspd" 

]

df_keep = df_final_imputed[keep_col]
df_keep.head()

Address    State County    City Date Local   NO2 Mean  \
638  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-01  17.716667   
639  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-02  15.062500   
640  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-03  21.643478   
641  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-04  25.041668   
642  400 W RIVER ROAD  Arizona   Pima  Tucson 2012-01-05  21.981817   

     NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI   O3 Mean  O3 1st Max Value  \
638               31.0                 0       29  0.013667             0.030   
639               30.6                18       28  0.015083             0.030   
640               31.0                18       29  0.011417             0.026   
641               37.8                10       35  0.009208             0.020   
642               37.1                17       35  0.013042             0.031   

     O3 1st Max Hour  O3 AQI  SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour  \
638               10      25  0.254167                0.5                19   
639               10      25  0.200000                0.6                19   
640                9      22  0.295455                0.7                 8   
641               10      17  0.737500                2.1                19   
642                9      26  0.330435                0.8                21   

     SO2 AQI   CO Mean  CO 1st Max Value  CO 1st Max Hour  CO AQI Population  \
638      0.0  0.336842               0.6                5     7.0    542,649   
639      0.0  0.225000               0.4               23     5.0    542,649   
640      0.0  0.295833               0.4                0     5.0    542,649   
641      3.0  0.345833               0.5               12     6.0    542,649   
642      0.0  0.291667               0.6               23     7.0    542,649   

     Latitude  Longitude  tmax  prcp  wspd  
638   31.9681  -111.7806  26.7   0.0  17.6  
639   31.9681  -111.7806  24.4   0.0  27.4  
640   31.9681  -111.7806  26.1   0.0  10.8  
641   31.9681  -111.7806  24.4   0.0   9.0  
642   31.9681  -111.7806  23.9   0.0   9.7

In [203]:
df_keep.shape

(58158, 27)

In [204]:
df_keep.to_csv("Dataset/EDA/pollution_us_2012_2016-population-weather.csv", index=False)

---

This will be the csv file with just the population, latitude and longitude columns added (132354 row). 

In [205]:
keep_col = ["Address",
            "State",
            "County",
            "City",
            "Date Local",
            "NO2 Mean",
            "NO2 1st Max Value",
            "NO2 1st Max Hour",
            "NO2 AQI",
            "O3 Mean",
            "O3 1st Max Value",
            "O3 1st Max Hour",
            "O3 AQI",
            "SO2 Mean",
            "SO2 1st Max Value",
            "SO2 1st Max Hour",
            "SO2 AQI",
            "CO Mean",
            "CO 1st Max Value", 
            "CO 1st Max Hour",
            "CO AQI",
            "Population",
            "Latitude",
            "Longitude",
]

df_keep = merged[keep_col]
print(df_keep.shape)
df_keep.head()

(132354, 24)


Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local   NO2 Mean  NO2 1st Max Value  NO2 1st Max Hour  NO2 AQI  \
0  2012-01-01  21.208332               33.0                 0       31   
1  2012-01-02  17.208332               38.0                22       36   
2  2012-01-03  30.000000               47.0                18       44   
3  2012-01-04  33.666668               47.0                19       44   
4  2012-01-05  31.695652               48.0                18       45   

    O3 Mean  O3 1st Max Value  O3 1st Max Hour  O3 AQI  SO2 Mean  \
0  0.015083             0.028               11      24  1.458333   
1  0.018042             0.034                9      29  0.416667   
2  0.008542             0.024               10      20  2.250000   
3  0.005458             0.016               10      14  2.791667   
4  0.008292             0.024                9      20  3.043478   

   SO2 1st Max Value  SO2 1st Max Hour  SO2 AQI   CO Mean  CO 1st Max Value  \
0                5.0                 0      7.0  1.152632               2.7   
1                2.0                 7      3.0  0.425000               0.5   
2                6.0                20      9.0  0.800000               1.7   
3                5.0                 7      7.0  1.275000               1.9   
4                7.0                22     10.0  1.045833               1.7   

   CO 1st Max Hour  CO AQI Population  Latitude  Longitude  
0                5    31.0  1,608,415   33.2798  -112.7681  
1                0     6.0  1,608,415   33.2798  -112.7681  
2               23    19.0  1,608,415   33.2798  -112.7681  
3                1    22.0  1,608,415   33.2798  -112.7681  
4               23    19.0  1,608,415   33.2798  -112.7681

In [206]:
df_keep.to_csv("Dataset/EDA/pollution_us_2012_2016-population.csv", index=False)

---

## Conclusions and Next Steps

- Columns for population, maximum temperature, precipitation and wind speed have been added
- This has resulted in the loss of 78,918 rows due to missing values
- A dataset with only population data added is a backup option
- The dataset is ready for EDA 